### PR TITLE
Issue2 models, forms

### PR DESCRIPTION
**Pull request by robbason**
I'll modify the Python version before merging

### CODE REVIEW

File: app/core/models.py [hunk 1]
```diff
@@ -1,3 +1,137 @@
...
+    end_at = models.DateTimeField()
+    repo_url = models.URLField()
+    # Data stored in S3 - privileges on S3 buckets will help
+    # prevent leakage of secret data
+    sample_data_url = models.URLField()
+    sample_score_reference_url = models.URLField()
+    secret_data_url = models.URLField()
+    secret_score_reference_url = models.URLField()
+    execution_options_json = models.JSONField()
+    # TODO: How to store scoring container reference in challenge?
```
Comment (mikemhenry) [inline app/core/models.py:31]: Can we associate an instance of our Container model?

File: app/core/views/submission.py [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+from django.urls import reverse_lazy
+from django.views.generic.edit import DeleteView
+from django.views.generic.detail import DetailView
+from django.views.generic.list import ListView
+
+from ..forms import SubmissionForm, ContainerForm
+from ..models import Submission
+
+
+class OwnerMatchMixin(LoginRequiredMixin, UserPassesTestMixin):
```
Comment (mikemhenry) [inline app/core/views/submission.py:15]: What does this class do?

File: app/core/views/submission.py [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+            container.save()
+            if submission_form.is_valid():
+                submission = submission_form.save(commit=False)
+                submission.container = container
+                submission.challenge = container.challenge
+                submission.user = request.user
+                submission.save()
+                return redirect("submission-detail", pk=submission.pk)
+
+    # context = {"container_form": container_form, "submission_form": submission_form}
```
Comment (mikemhenry) [inline app/core/views/submission.py:95]: What is the commented code out here for?

File: app/sampl/settings.py [hunk 4]
```diff
@@ -20,76 +21,78 @@
 # See https://docs.djangoproject.com/en/3.1/howto/deployment/checklist/
 
 # SECURITY WARNING: keep the secret key used in production secret!
-SECRET_KEY = '7&gz49!+qmxwltqo*!_g$n+i)%qcn9c^%2kbwzlnwiuofj29!%'
+SECRET_KEY = "7&gz49!+qmxwltqo*!_g$n+i)%qcn9c^%2kbwzlnwiuofj29!%"
```
Comment (mikemhenry) [inline app/sampl/settings.py:24]: Just going to make this issue so we remember to change this when we deploy (we should handle it like we do the DB connection)
https://github.com/samplchallenges/SAMPL-league/issues/20

File: app/sampl/settings_production.py [hunk 5]
```diff
@@ -0,0 +1,8 @@
+from .settings import *
+
```
Comment (mikemhenry) [inline app/sampl/settings_production.py:2]: Ah I see how this is setup -- should we do the DB connection the same way?

**Review by mikemhenry — CHANGES_REQUESTED** [review]
Pretty much just some questions -- we can talk about them on our call today -- looks great!

File: app/core/configurator.py [hunk 6]
```diff
@@ -0,0 +1,44 @@
...
+
+COMPUTING_AND_HARDWARE_DETAILS = """Please provide details of the computing resources that were used to train models and make predictions.
+Please specify compute time for training models and querying separately for empirical prediction methods.
+Provide a detailed description of the hardware used to run the simulations."""
+
+SOFTWARE_DETAILS = """List all major software packages used and their versions.
+Create a new line for each software."""
+
+CATEGORY_OPTIONS = ("Physical (MM)", "Physical (QM)", "Empirical", "Mixed")
+CATEGORY_CHOICES = zip(CATEGORY_OPTIONS, CATEGORY_OPTIONS)
```
Comment (mikemhenry) [inline app/core/configurator.py:23]: What does this do?

Comment (robbason) [reply app/core/configurator.py:23]: should be ```list(zip,``` thanks

Comment (robbason) [reply app/core/configurator.py:23]: django choices have to be a list of tuples, in this case I want both elements of the tuple to be the same

File: app/core/configurator.py [hunk 7]
```diff
@@ -0,0 +1,44 @@
...
+METHOD_DETAILS = """Methodology and computational details.
+Level of details should be roughly equivalent to that used in a publication.
+Please include the values of key parameters with units.
+Please explain how statistical uncertainties were estimated.
+
+If you have evaluated additional microstates, please report their SMILES strings and populations of all the microstates in this section.
+If you used a microstate other than the challenge provided microstate (`SMXX_micro000`), please list your chosen `Molecule ID` (in the form of `SMXX_extra001`) along with the SMILES string in your methods description.
+
+Use as many lines of text as you need."""
+# TODO: keep or remove?  All text following the 'Method:' keyword will be regarded as part of your free text methods description."""
```
[the remark this answers is not in the document]
Comment (robbason) [reply app/core/configurator.py:38]: removed

File: app/core/models.py (hunk 1 above)
Comment (robbason) [reply app/core/models.py:31]: One option:
- scoring_container - 'nullable'
- create the challenge
- create the container, associate to the challenge
- set the scoring container in the challenge

OR
- Container could be a top level, independent of Challenge
- Challenge has a nullable scoring_container fkey to container
- Relationship between submission, challenge, and container lives only in submission

```
             Scoring Container
                                 \
Container               Challenge
              \                /
               submission
```

File: app/core/views/submission.py (hunk 2 above)
Comment (robbason) [reply app/core/views/submission.py:15]: This mixin on a view will make sure the user accessing the object in the view is the owner of that object. test_func is the Django builtin that will be used in the standard Django object-based views

File: app/core/views/submission.py (hunk 3 above)
Comment (robbason) [reply app/core/views/submission.py:95]: removed

File: app/core/models.py (hunk 1 above)
Comment (robbason) [reply app/core/models.py:31]: I added a 'scoremaker' model to address this

File: app/sampl/settings.py (hunk 4 above)
Comment (robbason) [reply app/sampl/settings.py:24]: We discussed the solution, will be in update to PR

**Comment by mikemhenry** [comment]
just some notes when I tested this:
need to add `django-crispy-forms` to pipfile

**Comment by mikemhenry** [comment]
We get errors on these pages:
```
/admin/core/container/
no such column: core_container.registry

 /admin/core/container/
no such column: core_container.registry

/admin/core/prediction/
no such column: core_submissionevaluation.digest

/admin/core/prediction/add/
no such column: core_submissionevaluation.digest

/admin/core/prediction/
no such column: core_submissionevaluation.digest

/admin/core/prediction/add/
no such column: core_submissionevaluation.digest

/admin/core/scoremaker/
no such column: core_container.registry

/admin/core/scoremaker/add/
no such column: core_container.registry

 /admin/core/submissionevaluation/
no such column: core_submissionevaluation.digest

 /admin/core/submissionevaluation/add/
no such column: core_submission.draft_mode

 /admin/core/submission/
no such column: core_submission.draft_mode

/admin/core/submission/add/
no such column: core_container.registry
```

**Comment by mikemhenry** [comment]
removed comment about errors that were my fault! JSON stuff above still applies

**Review by mikemhenry — APPROVED** [review]

**Comment by robbason** [comment]
Still need to handle execution options better

**Pull request merged**